tools/ci/src/main.rs

```diff
@@ -0,0 +1,257 @@
+        run_bash($cmdline, $envs)
+    };
+}
+
+fn run_all_clap_subcommands(skips: &[String]) -> Result<()> {
+    let subcmds = Cli::command()
+        .get_subcommands()
+        .map(|sc| sc.get_name().to_string())
+        .collect::<Vec<_>>();
+
+    for subcmd in subcmds {
+        if skips.contains(&subcmd) {
+            log::info!("skipping {subcmd} as requested");
+            continue;
+        }
+        log::info!("executing cargo ci {subcmd}");
+        bash!(&format!("cargo ci {subcmd}"))?;
+    }
+
+    Ok(())
+}
+
+fn run_bash(cmdline: &str, additional_env: &[(&str, &str)]) -> Result<()> {
+    let mut env = env::vars().collect::<HashMap<_, _>>();
+    env.extend(additional_env.iter().map(|(k, v)| (k.to_string(), v.to_string())));
+    log::debug!("$ {cmdline}");
+    let status = cmd!("bash", "-lc", cmdline).full_env(env).run()?;
+    if !status.status.success() {
+        let e = anyhow::anyhow!("command failed: {cmdline}");
+        log::error!("{e}");
+        return Err(e);
+    }
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    let cli = Cli::parse();
+
+    match cli.cmd {
+        Some(CiCmd::Test) => {
+            bash!("cargo test --all -- --skip unreal")?;
+            // The fallocate tests have been flakely when running in parallel
+            bash!("cargo test -p spacetimedb-durability --features fallocate -- --test-threads=1")?;
+            bash!("bash tools/check-diff.sh")?;
+            bash!("cargo run -p spacetimedb-codegen --example regen-csharp-moduledef && bash tools/check-diff.sh crates/bindings-csharp")?;
+            bash!("(cd crates/bindings-csharp && dotnet test -warnaserror)")?;
+        }
+
+        Some(CiCmd::Lint) => {
+            bash!("cargo fmt --all -- --check")?;
+            bash!("cargo clippy --all --tests --benches -- -D warnings")?;
+            bash!("(cd crates/bindings-csharp && dotnet tool restore && dotnet csharpier --check .)")?;
+            // `bindings` is the only crate we care strongly about documenting,
+            // since we link to its docs.rs from our website.
+            // We won't pass `--no-deps`, though,
+            // since we want everything reachable through it to also work.
+            // This includes `sats` and `lib`.
+            bash!(
+                "cd crates/bindings && cargo doc",
+                // Make `cargo doc` exit with error on warnings, most notably broken links
+                &[("RUSTDOCFLAGS", "--deny warnings")]
+            )?;
+        }
+
+        Some(CiCmd::WasmBindings) => {
+            bash!("cargo test -p spacetimedb-codegen")?;
+            bash!("cargo update")?;
+            bash!("cargo run -p spacetimedb-cli -- build --project-path modules/module-test")?;
+        }
+
+        Some(CiCmd::Smoketests { args }) => {
+            // On some systems, there is no `python`, but there is `python3`.
+            let py3_available = cmd!("bash", "-lc", "command -v python3 >/dev/null 2>&1")
```

Comment:
I added this.. but I realize I'm not sure whether this works on windows.. Also I probably should have reused the `bash!` macro..